### DISCUSSION TITLE
Remark-math `$$` content on single-line is being considered inline math

### DISCUSSION THREAD

**Discussion by pjkaufman** [remarkjs/remark#1056] [Q&A] [answered]
I am trying to work on formatting some math blocks using remark to identify math blocks as opposed to inline math as the formatting I want to apply is different for each type. The problem is that the following seems to be considered inline math instead of block math:
``` math
$$L = \frac{1}{2} \rho v^2 S C_L$$
```

Here are two places that I have seen references stating that `$$` means math block and NOT inline math:
- https://github.com/remarkjs/remark/discussions/572
- https://docs.mathjax.org/en/latest/input/tex/delimiters.html#tex-delimiters

Now I could just be doing something wrong or have some kind of format misunderstanding, so I wanted to check here if that was the case.

Here is the AST parser result for the math provided above: https://astexplorer.net/#/gist/111c4f5728f9c80696a9422a7c9c99f0/aa4f9e212560494d7ebb5d826e6777137afa59d1

Here is also how I am reproducing the issue with a clean setup of the package:
`package.json`
``` json
{
  "name": "example",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "remark-gfm": "^3.0.1",
    "remark-math": "^5.1.1",
    "remark": "^14.0.2",
    "unified": "^10.1.2"
  }
}
```
`examples.md`
``` markdown
# Header
$$\boldsymbol{a}=\begin{bmatrix}a_x \\ a_y\end{bmatrix}$$
some more text

$$L = \frac{1}{2} \rho v^2 S C_L$$

```
`examples.js`
``` javascript
import fs from "node:fs/promises";
import {visit} from 'unist-util-visit';
import {remark} from 'remark';
import remarkMath from 'remark-math';
import remarkGfm from 'remark-gfm';

const file = await fs.readFile("example.md");
const fileContents = String(file);
const ast = remark().use(remarkGfm).use(remarkMath).parse(file);

console.log(ast)

visit(ast, 'math', (node) => {
  console.log(fileContents.substring(node.position.start.offset, node.position.end.offset) + '\nEND\n')

});

visit(ast, 'inlineMath', (node) => {
  console.log(fileContents.substring(node.position.start.offset, node.position.end.offset) + '\nEND2\n')
});
```

Here is the output I am getting:
``` text
{
  type: 'root',
  children: [
    {
      type: 'heading',
      depth: 1,
      children: [Array],
      position: [Object]
    },
    { type: 'paragraph', children: [Array], position: [Object] },
    { type: 'paragraph', children: [Array], position: [Object] }
  ],
  position: {
    start: { line: 1, column: 1, offset: 0 },
    end: { line: 6, column: 1, offset: 118 }
  }
}
$$\boldsymbol{a}=\begin{bmatrix}a_x \\ a_y\end{bmatrix}$$
END2

$$L = \frac{1}{2} \rho v^2 S C_L$$
END2
```

I am expecting
``` text
{
  type: 'root',
  children: [
    {
      type: 'heading',
      depth: 1,
      children: [Array],
      position: [Object]
    },
    { type: 'paragraph', children: [Array], position: [Object] },
    { type: 'paragraph', children: [Array], position: [Object] }
  ],
  position: {
    start: { line: 1, column: 1, offset: 0 },
    end: { line: 6, column: 1, offset: 118 }
  }
}
$$\boldsymbol{a}=\begin{bmatrix}a_x \\ a_y\end{bmatrix}$$
END1

$$L = \frac{1}{2} \rho v^2 S C_L$$
END1
```

I am on npm version `8.19.2` and node version `v16.16.0`.

Please let me know if this is how it is supposed to work. If that is the case, I can manage this by checking to see if the inline math block is surrounded by more than 1 `$`. But I would like to make sure this is not operator error on my end before I do so.

Thanks for the help!

**Comment by wooorm** [comment, marked as answer]
Math has no spec. Nobody knows what should happen. Not everyone agrees that what you want, is the way it should happen.

Particularly, in our case, we follow how code works in markdown, with two exceptions:
* dollars instead of backticks
* 2 is also fine for blocks
Code works like this:

````markdown
```inline code```

```
block code
```
````

For more information on how syntax works in our syntax plugins, see the Syntax section in each readme.
In this case: https://github.com/remarkjs/remark-math/tree/main/packages/remark-math#syntax

**Reply by pjkaufman** [reply]
Gotcha. I looked at the syntax, but never saw it say that a single and multiple dollar signs on a line would be mathInline. But what you say makes sense. I guess I just overlooked the statement somewhere. I will act accordingly with my code. Thanks!

**Reply by wooorm** [reply]
Yeah, I can understand that. Made it a bit more clear: https://github.com/remarkjs/remark-math/commit/f4596b4cddb85f89f2c40ec42b67e0f4995d6ead. Thanks!

**Reply by pjkaufman** [reply]
Thanks for that! I did add some comments for very minor grammar things and a question around wether it was the dollar sign count or the new line character that made the math a math block. Thanks again for these wonderful markdown packages!

**Reply by Abban-Fahim** [reply]
Sorry to disturb this thread, but do you guys have ideas on how to actually support this syntax (of double-dollar signs outputting display math). At one point there was a plugin option to enable it, spawned from remarkjs/remark-math#8, however since remark migrated to micromark, this option was removed.

Any possibility of re-implementing this option, or maybe I can find a way around this. I only really need a way of targeting "inline math" and "double-quote inline math" separately in CSS.

Thanks for any help!

**Reply by wooorm** [reply]
> I only really need a way of targeting "inline math" and "double-quote inline math" separately in CSS.

You can look at ASTs with plugins and change them.

It’s unclear to me what you mean with “inline math” and “double-quote inline math”. What about triple quotes? What about blocks? What is “display” even? Perhaps a new discussion with more info is more useful.

**Comment by 9xcode** [comment]
try `$$ formula $$` for Centered formula
and try `$ variable $` for inline text or formula 
ezzyy

**Comment by kimhanm** [comment]
> "Math has no spec. Nobody knows what should happen."

I have to disagree with this premise. The markdown specification for math is undoubtedly based on the $\TeX$ specification for math content.

In Donald Knuths "The TeXbook" Chapter 17 ("more about math"), he writes

> The normal way to typeset a formula with TEX is to enclose it in dollar signs `$ . . . $`; this yields the formula in text style (style T). 
> Or you can enclose it in double dollar signs `$$ . . . $$`; this displays the formula in display style (style D).

The same page even gives the example `$$1\over2$$`.
